### ISSUE TITLE
[package] qt/6.2.2: Component 'xkbcommon::libxkbcommon-x11' not found

### ISSUE THREAD

**Issue by cajus** [conan-io/conan-center-index#9123]
### Package and Environment Details (include every applicable attribute)
  * Package Name/Version: **qt/6.2.2**
  * Operating System+version: **Linux Fedora 35**
  * Compiler+version: **GCC 11**
  * Conan version: **conan 1.44.1*
  * Python version: **Python 3.10.2**

### Conan profile
```
[settings]
os=Linux
os_build=Linux
arch=x86_64
arch_build=x86_64
compiler=gcc
compiler.version=11
compiler.libcxx=libstdc++11
build_type=Release
[options]
OpenSSL:no_asm=True
OpenSSL:no_engine=True
OpenSSL:no_ssl2=False
OpenSSL:no_ssl3=True
OpenSSL:no_weak_ssl_ciphers=True
OpenSSL:shared=False
brotli:fPIC=True
brotli:shared=False
bzip2:shared=False
ffmpeg:alsa=True
freetype:shared=False
harfbuzz:shared=False
libcurl:shared=False
libdmtx:shared=False
libharu:shared=False
libjpeg:shared=False
libpng:shared=False
libsodium:shared=False
libwebp:fPIC=True
libwebp:shared=False
libxml2:fPIC=True
libxml2:iconv=False
libxml2:reader=False
libxml2:schemas=True
libxml2:shared=False
libxml2:writer=True
libxml2:xpath=True
mbedtls:fPIC=True
mbedtls:shared=False
mbedtls:with_zlib=True
pcre2:shared=False
pcre2:support_jit=False
qt:openssl=True
qt:qt3d=False
qt:qt5compat=True
qt:qtactiveqt=False
qt:qtcharts=True
qt:qtconnectivity=True
qt:qtdatavis3d=False
qt:qtdeclarative=True
qt:qtdoc=False
qt:qtimageformats=True
qt:qtlottie=False
qt:qtmultimedia=False
qt:qtnetworkauth=False
qt:qtremoteobjects=False
qt:qtscxml=False
qt:qtsensors=True
qt:qtserialbus=False
qt:qtserialport=False
qt:qtshadertools=True
qt:qtsvg=True
qt:qttools=True
qt:qttranslations=True
qt:qtvirtualkeyboard=False
qt:qtwayland=False
qt:qtwebchannel=False
qt:qtwebengine=False
qt:qtwebsockets=True
qt:qtwebview=False
qt:shared=False
qt:with_fontconfig=False
qt:with_glib=False
qt:with_harfbuzz=True
qt:with_icu=False
qt:with_jasper=False
qt:with_libalsa=False
qt:with_libpng=True
qt:with_mng=False
qt:with_mysql=False
qt:with_odbc=False
qt:with_openal=False
qt:with_pq=False
qt:with_sqlite3=False
qt:with_tiff=False
qt:with_webp=True
xkbcommon:shared=False
zxcvbn:shared=False
[conf]
[build_requires]
[env]
CFLAGS=-fPIC -D_FORTIFY_SOURCE=2 -Wformat-security -Wformat
CXXFLAGS=-fPIC -D_FORTIFY_SOURCE=2 -Wformat-security -Wformat
LDFLAGS=-Wl,--hash-style=both -Wl,-z,defs -Wl,-z,relro -fstack-protector-strong
```

### Steps to reproduce

I'm trying to build an application that uses Qt of the _conan-center-index_ Qt package recipe. It has been built locally with all required dependencies for a specific local channel _@foo/dev_.

These are relevant parts of the `conanfile.py` of the application:

```python
class AppConan(ConanFile):
    generators = ["CMakeToolchain", "cmake_find_package", "cmake_find_package_multi", "xcode"]

   def initialize(self, settings, env, buildenv=None):
      self.requires = ("openssl/1.1.1m@foo/dev", "libsodium/1.0.18@foo/dev", "libxml2/2.9.12@foo/dev", "zxcvbn/2.5@foo/dev", "libharu/2.3.0@foo/dev", "qt/6.2.2@foo/dev",)
```

Running `conan install` with the above profile results into

```
conanfile.py: ERROR: Generator cmake_find_package(file:None) failed
Component 'xkbcommon::libxkbcommon-x11' not found in 'xkbcommon' package requirement
ERROR: Component 'xkbcommon::libxkbcommon-x11' not found in 'xkbcommon' package requirement
```

and I'm wondering what it means to me. The xkbcommon package (latest from _conan-center-index_, also built for the _foo/dev_ channel) has been loaded by conan.

Is it because the built is configured to be "static"? Because of a non default channel?

### Logs (Include/Attach if Applicable)
<details><summary>Click to expand log</summary>

```
Building in /home/cajus/Projekte/bj/app/build

Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux
os_build=Linux
[options]
OpenSSL:no_asm=True
OpenSSL:no_engine=True
OpenSSL:no_ssl2=False
OpenSSL:no_ssl3=True
OpenSSL:no_weak_ssl_ciphers=True
OpenSSL:shared=False
brotli:fPIC=True
brotli:shared=False
bzip2:shared=False
ffmpeg:alsa=True
freetype:shared=False
harfbuzz:shared=False
libcurl:shared=False
libdmtx:shared=False
libharu:shared=False
libjpeg:shared=False
libpng:shared=False
libsodium:shared=False
libwebp:fPIC=True
libwebp:shared=False
libxml2:fPIC=True
libxml2:iconv=False
libxml2:reader=False
libxml2:schemas=True
libxml2:shared=False
libxml2:writer=True
libxml2:xpath=True
mbedtls:fPIC=True
mbedtls:shared=False
mbedtls:with_zlib=True
pcre2:shared=False
pcre2:support_jit=False
qt:openssl=True
qt:qt3d=False
qt:qt5compat=True
qt:qtactiveqt=False
qt:qtcharts=True
qt:qtconnectivity=True
qt:qtdatavis3d=False
qt:qtdeclarative=True
qt:qtdoc=False
qt:qtimageformats=True
qt:qtlottie=False
qt:qtmultimedia=False
qt:qtnetworkauth=False
qt:qtremoteobjects=False
qt:qtscxml=False
qt:qtsensors=True
qt:qtserialbus=False
qt:qtserialport=False
qt:qtshadertools=True
qt:qtsvg=True
qt:qttools=True
qt:qttranslations=True
qt:qtvirtualkeyboard=False
qt:qtwayland=False
qt:qtwebchannel=False
qt:qtwebengine=False
qt:qtwebsockets=True
qt:qtwebview=False
qt:shared=False
qt:with_fontconfig=False
qt:with_glib=False
qt:with_harfbuzz=True
qt:with_icu=False
qt:with_jasper=False
qt:with_libalsa=False
qt:with_libpng=True
qt:with_mng=False
qt:with_mysql=False
qt:with_odbc=False
qt:with_openal=False
qt:with_pq=False
qt:with_sqlite3=False
qt:with_tiff=False
qt:with_webp=True
xkbcommon:shared=False
zxcvbn:shared=False
[build_requires]
[env]
CFLAGS=-fPIC -D_FORTIFY_SOURCE=2 -Wformat-security -Wformat
CXXFLAGS=-fPIC -D_FORTIFY_SOURCE=2 -Wformat-security -Wformat
LDFLAGS=-Wl,--hash-style=both -Wl,-z,defs -Wl,-z,relro -fstack-protector-strong
Version ranges solved
    Version range '>=1.2.11' required by 'libharu/2.3.0@foo/dev' resolved to 'zlib/1.2.11@foo/dev' in local cache
    Version range '>=1.6.34' required by 'libharu/2.3.0@foo/dev' resolved to 'libpng/1.6.37@foo/dev' in local cache

conanfile.py: Installing package
Requirements
    OpenSSL/1.1.1m@foo/dev from 'gonicus-conan-repo' - Cache
    brotli/1.0.9@foo/dev from 'gonicus-conan-repo' - Cache
    bzip2/1.0.8@foo/dev from 'gonicus-conan-repo' - Cache
    double-conversion/3.1.5@foo/dev from 'gonicus-conan-repo' - Cache
    freetype/2.11.1@foo/dev from 'gonicus-conan-repo' - Cache
    gtest/1.11.0@foo/dev from 'gonicus-conan-repo' - Cache
    harfbuzz/3.2.0@foo/dev from 'gonicus-conan-repo' - Cache
    libcurl/7.81.0@foo/dev from 'gonicus-conan-repo' - Cache
    libharu/2.3.0@foo/dev from 'gonicus-conan-repo' - Cache
    libpng/1.6.37@foo/dev from 'gonicus-conan-repo' - Cache
    libsodium/1.0.18@foo/dev from 'gonicus-conan-repo' - Cache
    libxml2/2.9.12@foo/dev from local cache - Cache
    opengl/system@foo/dev from local cache - Cache
    openssl/1.1.1m@foo/dev from local cache - Cache
    pcre2/10.39@foo/dev from 'gonicus-conan-repo' - Cache
    qt/6.2.2@foo/dev from local cache - Cache
    xkbcommon/1.1.0@foo/dev from 'gonicus-conan-repo' - Cache
    xorg/system@foo/dev from local cache - Cache
    zlib/1.2.11@foo/dev from local cache - Cache
    zxcvbn/2.5@foo/dev from 'gonicus-conan-repo' - Cache
Packages
    OpenSSL/1.1.1m@foo/dev:8ee2d93e621f6fde2d8f504197969b893428af52 - Cache
    brotli/1.0.9@foo/dev:dfbe50feef7f3c6223a476cd5aeadb687084a646 - Cache
    bzip2/1.0.8@foo/dev:c32092bf4d4bb47cf962af898e02823f499b017e - Cache
    double-conversion/3.1.5@foo/dev:6557f18ca99c0b6a233f43db00e30efaa525e27e - Cache
    freetype/2.11.1@foo/dev:7d544953cc6cdd1bd06a4e40fc9904c650a37eb2 - Cache
    gtest/1.11.0@foo/dev:4fefe176018d6564b5e2144163a209c027be31c5 - Cache
    harfbuzz/3.2.0@foo/dev:e410fff732cb78b37c6cc39ac96da0e936c4b183 - Cache
    libcurl/7.81.0@foo/dev:d3d3b541e5fd3d131bbe37cdd021135e9b1c5583 - Cache
    libharu/2.3.0@foo/dev:b110308cc7ecbb92156a37f40d06071650d13f93 - Cache
    libpng/1.6.37@foo/dev:b88c434d8110e892fcfb7bbb4f76234f590822f1 - Cache
    libsodium/1.0.18@foo/dev:edb692f653d984500e90903ff82f78ba2ee02338 - Cache
    libxml2/2.9.12@foo/dev:26429fc704d7def17c76b2b527dfbde05b3b8ef7 - Cache
    opengl/system@foo/dev:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    openssl/1.1.1m@foo/dev:dfbe50feef7f3c6223a476cd5aeadb687084a646 - Cache
    pcre2/10.39@foo/dev:5ac5dc16910114fde903938811db7706bfb4361c - Cache
    qt/6.2.2@foo/dev:85fcf8409fbdc993288973fdd4813aae1dc7a6d2 - Cache
    xkbcommon/1.1.0@foo/dev:df30b38f0f776a68e249a593213aa07f3735d688 - Cache
    xorg/system@foo/dev:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    zlib/1.2.11@foo/dev:dfbe50feef7f3c6223a476cd5aeadb687084a646 - Cache
    zxcvbn/2.5@foo/dev:062863c92a5a0a247840166e9f84ebe8d10786b9 - Cache

Installing (downloading, building) binaries...
brotli/1.0.9@foo/dev: Already installed!
bzip2/1.0.8@foo/dev: Already installed!
double-conversion/3.1.5@foo/dev: Already installed!
gtest/1.11.0@foo/dev: Already installed!
libgsm/1.0.19@foo/dev: Already installed!
libsodium/1.0.18@foo/dev: Already installed!
libsrtp/2.4.2@foo/dev: Already installed!
opengl/system@foo/dev: Already installed!
openssl/1.1.1m@foo/dev: Already installed!
xkbcommon/1.1.0@foo/dev: Already installed!
xorg/system@foo/dev: Already installed!
zlib/1.2.11@foo/dev: Already installed!
zxcvbn/2.5@foo/dev: Already installed!
OpenSSL/1.1.1m@foo/dev: Already installed!
libpng/1.6.37@foo/dev: Already installed!
libxml2/2.9.12@foo/dev: Already installed!
libxml2/2.9.12@foo/dev: Appending PATH environment variable: /home/cajus/.conan/data/libxml2/2.9.12/foo/dev/package/26429fc704d7def17c76b2b527dfbde05b3b8ef7/bin
pcre2/10.39@foo/dev: Already installed!
freetype/2.11.1@foo/dev: Already installed!
libcurl/7.81.0@foo/dev: Already installed!
libharu/2.3.0@foo/dev: Already installed!
harfbuzz/3.2.0@foo/dev: Already installed!
qt/6.2.2@foo/dev: Already installed!
conanfile.py: ERROR: Generator cmake_find_package(file:None) failed
Component 'xkbcommon::libxkbcommon-x11' not found in 'xkbcommon' package requirement
ERROR: Component 'xkbcommon::libxkbcommon-x11' not found in 'xkbcommon' package requirement
```
</details>

**Comment by ericLemanissier** [comment]
according to https://github.com/conan-io/conan-center-index/blob/master/recipes/xkbcommon/all/conanfile.py#L136, the component `libxkbcommon-x11` is always created, if option `xkbcommon::with_x11` is True, so I can see 3 explanations to this error:

1. xkbcommon::with_x11 is set to false. No recipe on CCI does this, so maybe it is done in one of your recipes/your profile/your command line ?
2. You are not using [this recipe](https://github.com/conan-io/conan-center-index/blob/master/recipes/xkbcommon/all/conanfile.py) for xkbcommon
3. There is a bug in conan

**Comment by cajus** [comment]
Thanks for the quick answer @ericLemanissier,

for 1: as you can see above, there's no `xkbcommon:with_x11=false` setting in the profile above. If I add some debug output to the `with_x11` branch mentioned above, it runs into it.

for 2: that's copied, but there's a small diff, as I was unable to find out how to set a default channel for `conan create` yet:

```diff
diff /home/cajus/tmp/conan-center-index/recipes/xkbcommon/all/conanfile.py ./conanfile.py
60c61
<         self.requires("xorg/system")
---
>         self.requires("xorg/system@foo/dev")
62c63
<             self.requires("libxml2/2.9.12")
---
>             self.requires("libxml2/2.9.12@foo/dev")
64,65c65,66
<             self.requires("wayland/1.20.0")
<             self.requires("wayland-protocols/1.24")  # FIXME: This should be a build-requires
---
>             self.requires("wayland/1.20.0@foo/dev")
>             self.requires("wayland-protocols/1.24@foo/dev")  # FIXME: This should be a build-requires
72,73c73,74
<         self.build_requires("meson/0.60.2")
<         self.build_requires("bison/3.7.6")
---
>         self.build_requires("meson/0.60.2@foo/dev")
>         self.build_requires("bison/3.7.6@foo/dev")
75c76
<             self.build_requires("wayland/1.20.0")
---
>             self.build_requires("wayland/1.20.0@foo/dev")
```

for 3: if 1/2 are the only possibilities, I can move the issue to the conan project.

**Comment by ericLemanissier** [comment]
Please make a reproducible exemple based only on recipes available on CCI.
Also, what is this `initialize` function you are calling ? is it a documented conan feature ?

**Comment by cajus** [comment]
`initialize` is a member function of `ConanFile` which is just overridden in my case. I guess you can ignore it. I've attached a simple example with a non dynamic `conanfile.txt` file.

[qt-test.tar.gz](https://github.com/conan-io/conan-center-index/files/7958242/qt-test.tar.gz)

```
$ conan install --profile profiles/linux.gcc.11 .                                                                                                                               ✔ 
Configuration:
[settings]
arch=x86_64
arch_build=x86_64
build_type=Release
compiler=gcc
compiler.libcxx=libstdc++11
compiler.version=11
os=Linux
os_build=Linux
[options]
OpenSSL:no_asm=True
OpenSSL:no_engine=True
OpenSSL:no_ssl2=False
OpenSSL:no_ssl3=True
OpenSSL:no_weak_ssl_ciphers=True
OpenSSL:shared=False
brotli:fPIC=True
brotli:shared=False
bzip2:shared=False
ffmpeg:alsa=True
freetype:shared=False
harfbuzz:shared=False
libcurl:shared=False
libdmtx:shared=False
libharu:shared=False
libjpeg:shared=False
libpng:shared=False
libsodium:shared=False
libwebp:fPIC=True
libwebp:shared=False
libxml2:fPIC=True
libxml2:reader=False
libxml2:schemas=True
libxml2:shared=False
libxml2:writer=True
libxml2:xpath=True
mbedtls:fPIC=True
mbedtls:shared=False
mbedtls:with_zlib=True
pcre2:shared=False
pcre2:support_jit=False
qt:openssl=True
qt:qt3d=False
qt:qt5compat=True
qt:qtactiveqt=False
qt:qtcharts=True
qt:qtconnectivity=True
qt:qtdatavis3d=False
qt:qtdeclarative=True
qt:qtdoc=False
qt:qtimageformats=True
qt:qtlottie=False
qt:qtmultimedia=False
qt:qtnetworkauth=False
qt:qtremoteobjects=False
qt:qtscxml=False
qt:qtsensors=True
qt:qtserialbus=False
qt:qtserialport=False
qt:qtshadertools=True
qt:qtsvg=True
qt:qttools=True
qt:qttranslations=True
qt:qtvirtualkeyboard=False
qt:qtwayland=False
qt:qtwebchannel=False
qt:qtwebengine=False
qt:qtwebsockets=True
qt:qtwebview=False
qt:shared=False
qt:with_fontconfig=False
qt:with_glib=False
qt:with_harfbuzz=True
qt:with_icu=False
qt:with_jasper=False
qt:with_libalsa=False
qt:with_libpng=True
qt:with_mng=False
qt:with_mysql=False
qt:with_odbc=False
qt:with_openal=False
qt:with_pq=False
qt:with_sqlite3=False
qt:with_tiff=False
qt:with_webp=True
xkbcommon:shared=False
zxcvbn:shared=False
[build_requires]
[env]
CFLAGS=-fPIC -D_FORTIFY_SOURCE=2 -Wformat-security -Wformat
CXXFLAGS=-fPIC -D_FORTIFY_SOURCE=2 -Wformat-security -Wformat
LDFLAGS=-Wl,--hash-style=both -Wl,-z,defs -Wl,-z,relro -fstack-protector-strong
conanfile.txt: Installing package
Requirements
    brotli/1.0.9@foo/dev from 'gonicus-conan-repo' - Cache
    bzip2/1.0.8@foo/dev from 'gonicus-conan-repo' - Cache
    double-conversion/3.1.5@foo/dev from 'gonicus-conan-repo' - Cache
    freetype/2.11.1@foo/dev from 'gonicus-conan-repo' - Cache
    harfbuzz/3.2.0@foo/dev from 'gonicus-conan-repo' - Cache
    libpng/1.6.37@foo/dev from 'gonicus-conan-repo' - Cache
    opengl/system@foo/dev from local cache - Cache
    openssl/1.1.1m@foo/dev from local cache - Cache
    pcre2/10.39@foo/dev from 'gonicus-conan-repo' - Cache
    qt/6.2.2@foo/dev from local cache - Cache
    xkbcommon/1.1.0@foo/dev from 'gonicus-conan-repo' - Cache
    xorg/system@foo/dev from local cache - Cache
    zlib/1.2.11@foo/dev from local cache - Cache
Packages
    brotli/1.0.9@foo/dev:dfbe50feef7f3c6223a476cd5aeadb687084a646 - Cache
    bzip2/1.0.8@foo/dev:c32092bf4d4bb47cf962af898e02823f499b017e - Cache
    double-conversion/3.1.5@foo/dev:6557f18ca99c0b6a233f43db00e30efaa525e27e - Cache
    freetype/2.11.1@foo/dev:7d544953cc6cdd1bd06a4e40fc9904c650a37eb2 - Cache
    harfbuzz/3.2.0@foo/dev:e410fff732cb78b37c6cc39ac96da0e936c4b183 - Cache
    libpng/1.6.37@foo/dev:b88c434d8110e892fcfb7bbb4f76234f590822f1 - Cache
    opengl/system@foo/dev:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    openssl/1.1.1m@foo/dev:dfbe50feef7f3c6223a476cd5aeadb687084a646 - Cache
    pcre2/10.39@foo/dev:5ac5dc16910114fde903938811db7706bfb4361c - Cache
    qt/6.2.2@foo/dev:85fcf8409fbdc993288973fdd4813aae1dc7a6d2 - Cache
    xkbcommon/1.1.0@foo/dev:df30b38f0f776a68e249a593213aa07f3735d688 - Cache
    xorg/system@foo/dev:5ab84d6acfe1f23c4fae0ab88f26e3a396351ac9 - Cache
    zlib/1.2.11@foo/dev:dfbe50feef7f3c6223a476cd5aeadb687084a646 - Cache

Installing (downloading, building) binaries...
brotli/1.0.9@foo/dev: Already installed!
bzip2/1.0.8@foo/dev: Already installed!
double-conversion/3.1.5@foo/dev: Already installed!
opengl/system@foo/dev: Already installed!
openssl/1.1.1m@foo/dev: Already installed!
xkbcommon/1.1.0@foo/dev: Already installed!
xorg/system@foo/dev: Already installed!
zlib/1.2.11@foo/dev: Already installed!
libpng/1.6.37@foo/dev: Already installed!
pcre2/10.39@foo/dev: Already installed!
freetype/2.11.1@foo/dev: Already installed!
harfbuzz/3.2.0@foo/dev: Already installed!
qt/6.2.2@foo/dev: Already installed!
conanfile.txt: ERROR: Generator cmake_find_package_multi(file:None) failed
Component 'xkbcommon::libxkbcommon-x11' not found in 'xkbcommon' package requirement
ERROR: Component 'xkbcommon::libxkbcommon-x11' not found in 'xkbcommon' package requirement
```
You need to build the packages for the `foo@dev` channel, though.

**Comment by ericLemanissier** [comment]
if it's not documented, don't use it. it could very well be the cause of your error.
I tried your test case, but it failed with `ERROR: Unable to find 'qt/6.2.2@foo/dev' in remotes` so I removed `@foo/dev` from conanfile.txt and now i get `ERROR: qt/6.2.2: option 'with_jasper' doesn't exist`, so you don't seem to be using the qt recipe from conan-center

**Comment by cajus** [comment]
Yes, sorry. I needed four more flags, but 100% sure that these modifications will not affect what I see here. Nevertheless, I've updated the profile and reverted the package to it's conan-center-index one, except this:

```diff
         if not self.options.openssl:
             self._cmake.definitions["INPUT_openssl"] = "no"
         else:
-            if self.options["openssl"].shared:
-                self._cmake.definitions["INPUT_openssl"] = "runtime"
-            else:
-                self._cmake.definitions["INPUT_openssl"] = "linked"
+            self._cmake.definitions["INPUT_openssl"] = "linked"
```

There's no `.shared` available on that object for whatever reason, so I've removed it for now, in order to be able to build.

Here's the `base` profile without the four additional flags: [profiles.tar.gz](https://github.com/conan-io/conan-center-index/files/7958673/profiles.tar.gz)

I'm getting the same error. Will try without the `@foo/dev`, but as there's no matching build configuration that will take some time.

**Comment by ericLemanissier** [comment]
it works on my machine:
[output.txt](https://github.com/conan-io/conan-center-index/files/7959149/output.txt)

The packages listed at the beginning of your log, after `Requirements` is very different from mine, so there must be several recipes which are different. Please try to reproduce your problem with the actual packages currently in conan-center, or there is no way we can help you.

**Comment by cajus** [comment]
@ericLemanissier if you say "are verry different from yours" - I needed to add the channel name for all of the dependencies, because there seems to be no way to set a default channel for `requires()` resolution right now -> conan-io/conan/issues/10441
. If you know a way to do it, you're welcome ;-) 

So they're up to date. Mostly.

What *was* different was the _xkbcommon_ package, which had version _1.1.x_ instead of _1.3.x_, actually - maybe because I've synced that some time ago and forgot to update it. It finally works with a custom channel using the latest `xkbcommon` package.

The lazy _openssl_ modification from above is still need though. Will inspect it next week...

Sorry, my fault - thanks again for your time.
